xlators/protocol/server/src/server.c

```diff
@@ -627,12 +682,41 @@ server_graph_janitor_threads(void *data)
                " %s stack",
                victim->name);
         xlator_mem_cleanup(victim);
+        LOCK(&ctx->volfile_lock);
+        {
+            totchildcount = 0;
+            for (trav_p = &top->children; *trav_p; trav_p = &(*trav_p)->next) {
+                totchildcount++;
+                break;
+            }
+            if (!totchildcount && !ctx->destroy_ctx) {
+                ctx->destroy_ctx = _gf_true;
+                destroy_ctx = _gf_true;
+            }
+        }
+        UNLOCK(&ctx->volfile_lock);
         rpcsvc_autoscale_threads(ctx, conf->rpc, -1);
     }
 
 out:
     free(arg->victim_name);
     free(arg);
+    if (!totchildcount && destroy_ctx) {
+        gf_log(THIS->name, GF_LOG_INFO,
+               "Going to Cleanup ctx pool memory and exit the process %s",
+               ctx->cmdlinestr);
+        syncenv_destroy(ctx->env);
+        ctx->env = NULL;
+        gf_event_dispatch_destroy(ctx->event_pool);
+        if (ctx->mgmt) {
+            rpc = ctx->mgmt;
+            rpc_clnt_connection_cleanup(&rpc->conn);
+            rpc_clnt_unref(rpc);
+        }
+        server_cleanup(this, conf);
+        glusterfs_ctx_pool_destroy(ctx);
+        exit(0);
```

Comment:
If we do a complete and proper shutdown sequence, why do we need to explicitly call `exit()` here ? we could let the thread to terminate and the main program to finish normally (returning from `main()`).